File: UICollectionViewDataSourcePrefetching/UICollectionViewDataSourcePrefetchingObjCSample/UICollectionViewDataSourcePrefetchingObjCSample/MainCollectionViewController.m
//
//  MainCollectionViewController.m
//  UICollectionViewDataSourcePrefetchingObjCSample
//
//  Created by Nikolay Andonov on 10/19/16.
//  Copyright © 2016 Mentormate. All rights reserved.
//

#import "MainCollectionViewController.h"
#import "MainCollectionViewCell.h"

static CGFloat const CollectionViewCellPadding = 10;
static NSInteger const NumberOfCellsPerRow = 2;

@interface MainCollectionViewController () <UICollectionViewDataSourcePrefetching>

@property (assign, nonatomic) CGFloat collectionViewCellSize;

@property (strong, nonatomic) NSMutableArray<NSURL *> *imageURLs;
@property (strong, nonatomic) NSOperationQueue *downloadImageOperationQueue;
@property (strong, nonatomic) NSMutableDictionary<NSURL *, NSBlockOperation *> *operations;
@property (strong, nonatomic) NSMutableDictionary<NSURL *, UIImage *> *images;


@end

@implementation MainCollectionViewController


- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self configureCollectionView];
    [self populateModels];
    
    [self.collectionView reloadData];
}

- (void)configureCollectionView {
    
    CGFloat screenWidth = CGRectGetWidth(self.view.frame);
    CGFloat cellsAreaOnSingleRow = screenWidth - ((NumberOfCellsPerRow + 1) * CollectionViewCellPadding);
    self.collectionViewCellSize = cellsAreaOnSingleRow / NumberOfCellsPerRow;
    
    self.collectionView.contentInset = UIEdgeInsetsMake(CollectionViewCellPadding, CollectionViewCellPadding, CollectionViewCellPadding, CollectionViewCellPadding);
    
    self.collectionView.prefetchDataSource = self;
}

- (void)populateModels {
    
    self.downloadImageOperationQueue = [[NSOperationQueue alloc] init];

    self.imageURLs = [NSMutableArray array];
    self.operations = [NSMutableDictionary dictionary];
    self.images = [NSMutableDictionary dictionary];
    
    //Simulating initial load of content
    for (NSInteger counter = 0; counter < 50; counter++) {
        
        //Simulating slow download using large images
        NSString *imageStringAdress = [NSString stringWithFormat:@"http://placehold.it/2000x2000&text=SampleImage%ld",(long)(counter + 1)];
        NSURL *imageURL = [NSURL URLWithString:imageStringAdress];
        [self.imageURLs addObject:imageURL];
    }
}



#pragma mark <UICollectionViewDataSource>

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    
    return 1;
}


- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    
    return self.imageURLs.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    MainCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"Cell" forIndexPath:indexPath];
    NSURL *imageURL = self.imageURLs[indexPath.row];
    if (self.images[imageURL]) {
        cell.imageView.image = self.images[imageURL];
        [cell.activityIndicator stopAnimating];
    }
    else {
        [self executeDownloadImageOperationBlockForIndexPath:indexPath];
    }
    
    return cell;
}

#pragma mark <UICollectionViewDelegate>

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    return CGSizeMake(self.collectionViewCellSize, self.collectionViewCellSize);
}

- (void)collectionView:(UICollectionView *)collectionView didEndDisplayingCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath {
    
    [self cancelDowloandImageOperationBlockForIndexPath:indexPath];
}


#pragma mark <UICollectionViewDataSourcePrefetching>

// indexPaths are ordered ascending by geometric distance from the collection view
- (void)collectionView:(UICollectionView *)collectionView prefetchItemsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths {
    
    for (NSIndexPath *indexPath in indexPaths) {
        
        // Updating upcoming CollectionView's data source. Not assiging any direct value
        // as this operation is expensive it is performed on a private queue
        NSURL *imageURL = self.imageURLs[indexPath.row];
        if (!self.images[imageURL]) {
            [self executeDownloadImageOperationBlockForIndexPath:indexPath];
            NSLog(@"Prefetching data for indexPath: %@", indexPath);
        }
    }
}

// indexPaths that previously were considered as candidates for pre-fetching, but were not actually used; may be a subset of the previous call to -collectionView:prefetchItemsAtIndexPaths:
- (void)collectionView:(UICollectionView *)collectionView cancelPrefetchingForItemsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths {
    
    for (NSIndexPath *indexPath in indexPaths) {
        
        //Unloading or data load operation cancellations should happend here
        NSURL *imageURL = self.imageURLs[indexPath.row];
        if (self.operations[imageURL]) {
            [self cancelDowloandImageOperationBlockForIndexPath:indexPath];
            NSLog(@"Unloading data fetch in progress for indexPath: %@", indexPath);
        }
    }
    
}

#pragma mark - Utilities

- (void)executeDownloadImageOperationBlockForIndexPath:(NSIndexPath *)indexPath {
    
    NSURL *url = self.imageURLs[indexPath.row];
    NSBlockOperation *blockOperation = [[NSBlockOperation alloc] init];
    __weak NSBlockOperation *weakBlockOperation = blockOperation;
    __weak typeof(self) weakSelf = self;
    
    [blockOperation addExecutionBlock:^{
        if (weakBlockOperation.isCancelled) {
            weakSelf.operations[url] = nil;
            return;
        }
        
        NSData *imageData = [NSData dataWithContentsOfURL:url];
        UIImage *image = [UIImage imageWithData:imageData];
        weakSelf.images [url] = image;
        weakSelf.operations[url] = nil;
        
        NSArray *visibleCellIndexPaths = [weakSelf.collectionView indexPathsForVisibleItems];
        if ([visibleCellIndexPaths containsObject:indexPath]) {
            
            MainCollectionViewCell *cell = (MainCollectionViewCell *)[weakSelf.collectionView cellForItemAtIndexPath:indexPath];
            dispatch_async(dispatch_get_main_queue(), ^{
                cell.imageView.image = image;
                [cell.activityIndicator stopAnimating];
            });
        }
    }];
    [self.downloadImageOperationQueue addOperation:blockOperation];
    self.operations[url] = blockOperation;
    
}

- (void)cancelDowloandImageOperationBlockForIndexPath:(NSIndexPath *)indexPath {
    
    NSURL *imageURL = self.imageURLs[indexPath.row];
    if (self.operations[imageURL]) {
        NSBlockOperation *blockOperation = self.operations[imageURL];
        [blockOperation cancel];
        self.operations[imageURL] = nil;
    }
}

@end

File: UICollectionViewDataSourcePrefetching/UICollectionViewDataSourcePrefetchingSwiftSample/UICollectionViewDataSourcePrefetchingSwiftSample/MainCollectionViewController.swift
//
//  MainCollectionViewController.swift
//  UICollectionViewDataSourcePrefetchingSwiftSample
//
//  Created by Nikolay Andonov on 10/24/16.
//  Copyright © 2016 Mentormate. All rights reserved.
//

import UIKit

let CollectionViewCellPadding : CGFloat = 10.0
let NumberOfCellsPerRow : NSInteger = 2

class MainCollectionViewController: UICollectionViewController, UICollectionViewDataSourcePrefetching, UICollectionViewDelegateFlowLayout {
    
    var collectionViewCellSize = 0.0
    
    var imageURLs : [URL] = []
    var downloadImageOperationQueue = OperationQueue()
    var operations : [URL : BlockOperation] = [:]
    var images : [URL : UIImage] = [:]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        configureCollectionView()
        populateModel()
        collectionView?.reloadData()
    }


    func configureCollectionView() {
        
        let screenWidth = view.frame.width;
        let cellsAreaOnSingleRow = screenWidth - (CGFloat(NumberOfCellsPerRow + 1) * CollectionViewCellPadding)
        collectionViewCellSize = Double(cellsAreaOnSingleRow) / Double(NumberOfCellsPerRow);
        collectionView?.contentInset = UIEdgeInsetsMake(CollectionViewCellPadding, CollectionViewCellPadding, CollectionViewCellPadding, CollectionViewCellPadding);
        collectionView?.prefetchDataSource = self
        collectionView?.collectionViewLayout = UICollectionViewFlowLayout()
    }
    
    func populateModel() {
        
        //Simulating initial load of content
        for counter  in 0..<50 {
            
            //Simulating slow download using large images
            let imageStringAdress = "http://placehold.it/2000x2000&text=SampleImage\(counter + 1)"
            if let imageURL = URL.init(string: imageStringAdress) {
                imageURLs.append(imageURL)
            }
        }
    }
    
    // MARK: UICollectionViewDataSource

    override func numberOfSections(in collectionView: UICollectionView) -> Int {
        
        return 1
    }


    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        
        return imageURLs.count
    }

    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        
        let cell : MainCollectionViewCell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath) as! MainCollectionViewCell
        let url = imageURLs[indexPath.row]
        if images[url] != nil {
            cell.imageView.image = images[url]
            cell.activityIndicator.stopAnimating()
        }
        else {
            executeDownloadImageOperationBlock(indexPath: indexPath)
        }
        
        return cell
    }

    // MARK: UICollectionViewDelegate
    
    func collectionView(_ collectionView : UICollectionView,layout collectionViewLayout:UICollectionViewLayout,sizeForItemAt indexPath:IndexPath) -> CGSize {
        
        let cellSize:CGSize = CGSize(width: collectionViewCellSize, height: collectionViewCellSize)
        return cellSize
    }
    
    override func collectionView(_ collectionView: UICollectionView, didEndDisplaying cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
        
        cancelDowloandImageOperationBlock(indexPath: indexPath)
    }
    
    // MARK: UICollectionViewDataSourcePrefething
    
    // indexPaths are ordered ascending by geometric distance from the collection view
    func collectionView(_ collectionView: UICollectionView, prefetchItemsAt indexPaths: [IndexPath]) {
        
        for indexPath in indexPaths {
            
            // Updating upcoming CollectionView's data source. Not assiging any direct value
            // as this operation is expensive it is performed on a private queue
            let url = imageURLs[indexPath.row]
            if images[url] == nil {
                executeDownloadImageOperationBlock(indexPath: indexPath)
                print("Prefetching data for indexPath: \(indexPath)")
            }
        }
    }
    
    // indexPaths that previously were considered as candidates for pre-fetching, but were not actually used; may be a subset of the previous call to -collectionView:prefetchItemsAtIndexPaths:
    func collectionView(_ collectionView: UICollectionView, cancelPrefetchingForItemsAt indexPaths: [IndexPath]) {
        
        for indexPath in indexPaths {
            
            //Unloading or data load operation cancellations should happend here
            let url = imageURLs[indexPath.row]
            if operations[url] != nil {
                cancelDowloandImageOperationBlock(indexPath: indexPath)
                print("Unloading data fetch in progress for indexPath: \(indexPath)")
            }
        }
    }

    // MARK: Utilitites
    
    private func executeDownloadImageOperationBlock(indexPath: IndexPath) {
        
        let url = imageURLs[indexPath.row]
        let blockOperation = BlockOperation()
        weak var weakBlockOperation = blockOperation
        weak var weakSelf = self
        
        blockOperation .addExecutionBlock {
            
            if (weakBlockOperation?.isCancelled)! {
                weakSelf?.operations[url] = nil
                return;
            }
            
            var image : UIImage? = nil
            do {
                
                let imageData = try Data.init(contentsOf: url)
                image = UIImage.init(data: imageData)
                weakSelf?.images [url] = image
                weakSelf?.operations[url] = nil
            } catch {
            }
            
            let visibleCellIndexPaths = weakSelf?.collectionView?.indexPathsForVisibleItems
            if (visibleCellIndexPaths?.contains(indexPath))! {
                
                let cell : MainCollectionViewCell = weakSelf?.collectionView?.cellForItem(at: indexPath) as! MainCollectionViewCell
                
                DispatchQueue.main.async {
                    cell.imageView.image = image
                    cell.activityIndicator.stopAnimating()
                }
            }
        }
        downloadImageOperationQueue.addOperation(blockOperation)
        operations[url] = blockOperation
    }

    private func cancelDowloandImageOperationBlock(indexPath: IndexPath){
        
        let url = imageURLs[indexPath.row]
        if operations[url] != nil {
            let blockOperation = operations[url]
            blockOperation?.cancel()
            operations[url] = nil
        }
    }
}
